add: expense tracker tutorial

GitBolt opened the pull request:
Tutorial for Expense Tracker using PDAs for the upcoming Solana Workshop https://github.com/Solana-Workshops/.contributing/issues/2

acheroncrypto commented on client/src/tutorials/ExpenseTracker/files/index.test.ts.raw:
Have you tested this in playground?

acheroncrypto commented on client/src/tutorials/ExpenseTracker/files/lib.rs:
Could you format this file with rustfmt?

acheroncrypto commented on client/src/tutorials/ExpenseTracker/files/lib.rs:
```diff
@@ -0,0 +1,138 @@
+            + (4 + 12) // merchant name
+            + 8 // amount
+            + 1, // bump
+        seeds = [b"expense".as_ref(),authority.key().as_ref(),id.as_ref()], 
```
Byte string `b"expense"` is already a reference, `.as_ref()` is redundant, and some spacing would help.
```suggestion
        seeds = [b"expense", authority.key().as_ref(), id.as_ref()], 
```

acheroncrypto commented on client/src/tutorials/ExpenseTracker/files/lib.rs:
```diff
@@ -0,0 +1,138 @@
+    pub expense_account: Account<'info, ExpenseAccount>,
+
+    // Misc Accounts
+    #[account(address = system_program::ID)]
```
Redundant check, `pub system_program: Program<'info,System>,` already does this check.
```suggestion
```

acheroncrypto commented on client/src/tutorials/ExpenseTracker/files/lib.rs:
```diff
@@ -0,0 +1,138 @@
+    #[account(address = system_program::ID)]
+    pub system_program: Program<'info,System>,
+    #[account(address = solana_program::sysvar::rent::ID)]
+    pub rent: Sysvar<'info, Rent>,
```
Rent is not necessary to pass as an account, you can get it with `Rent::get()?` inside the program which decreases transaction size and is potentially more secure.

acheroncrypto commented on client/src/tutorials/ExpenseTracker/files/lib.rs:
```diff
@@ -0,0 +1,138 @@
+        expense_account.mname = mname;
+        expense_account.amount = amount;
+        expense_account.owner = *ctx.accounts.authority.key;
+        expense_account.bump = *ctx.bumps.get("expense_account").unwrap();
```
Saving bumps is not necessary, using `#[account(bump, ...)]` will get the same result.

acheroncrypto reviewed (changes requested):
Thank you. Could you please format the code with `rustfmt` and generally follow Rust's guidelines? It's much better to have a standardized format for tutorials.

I've also noted some of the redundant checks and please make sure to test this and confirm it works as expected in playground.

GitBolt commented:
Appreciate the detailed review! Fixing these and getting back you

GitBolt commented:
hey @acheroncrypto, i've made the changes. please have a look! i've fixed tests, formatted code, removed the unnecessary stuff and checked everything by running in playground

acheroncrypto commented on client/src/tutorials/ExpenseTracker/files/lib.rs:
```diff
@@ -0,0 +1,107 @@
+
+    pub fn initialize_expense(
+        ctx: Context<InitializeExpense>,
+        id: String,
```
Why pass a `String` here if the program panics for values that are not parseable to `u64`?

```suggestion
        id: u64,
```

GitBolt replied:
I use string cause the id is seed as well and it's easier to process strings

acheroncrypto replied:
I don't think it's justified. Clients should adjust based on the on-chain program, not the other way around.

GitBolt replied:
Updated, can you check now?

acheroncrypto commented on client/src/tutorials/ExpenseTracker/files/lib.rs:
```diff
@@ -0,0 +1,107 @@
+        amount: u64,
+    ) -> Result<()> {
+        let expense_account = &mut ctx.accounts.expense_account;
+        let integer_id = id.parse::<u64>().unwrap();
+        let string_id = integer_id.to_string();
+
+        require_eq!(string_id, id);
```
We are parsing the `String` to `u64` and then we are converting the parsed `u64` value to `String`. Lastly, we compare the two `String` values to be equal.

```suggestion
```

GitBolt replied:
Removed

acheroncrypto commented on client/src/tutorials/ExpenseTracker/files/index.test.ts.raw:
```diff
@@ -0,0 +1,45 @@
+describe("Expense Tracker", async () => {
+  let merchant_name = "test";
```
Can we use camelCase naming in TypeScript files for consistency?

acheroncrypto commented on client/src/tutorials/ExpenseTracker/pages/5.md:
````diff
@@ -0,0 +1,48 @@
+Have a look into this snippet:
+
+```ts
+const expenses = await program.account.expenseAccount.all([
````
```suggestion
const expenses = await pg.program.account.expenseAccount.all([
```

acheroncrypto commented on client/src/tutorials/ExpenseTracker/pages/5.md:
````diff
@@ -0,0 +1,48 @@
+Turns out, it's super easy!
+Have a look into this snippet:
+
+```ts
````
Can we format this snippet?

acheroncrypto commented on client/src/tutorials/ExpenseTracker/files/lib.rs:
```diff
@@ -0,0 +1,107 @@
+}
+
+#[derive(Accounts)]
+#[instruction(id : String)]
```
```suggestion
#[instruction(id: String)]
```

acheroncrypto reviewed (commented):
Thank you for fixing the earlier review comments, it looks much better now.

GitBolt commented:
I've fixed other issues, do let me know your thoughts on the string/u64 one in the instruction

acheroncrypto commented on client/src/tutorials/ExpenseTracker/pages/5.md:
```diff
@@ -0,0 +1,45 @@
+  {
+    memcmp: {
+      offset: 8 + 8,
+      bytes: wallet.publicKey.toBase58(),
```
```suggestion
      bytes: pg.wallet.publicKey.toBase58(),
```

acheroncrypto commented:
Looking good overall but CI is failing due to formatting.

GitBolt commented:
Oh weird, I did format the code through Solpg. Let me check again

acheroncrypto commented:
> Oh weird, I did format the code through Solpg. Let me check again

You can run `yarn format` to fix it.

GitBolt commented:
> > Oh weird, I did format the code through Solpg. Let me check again
> 
> You can run `yarn format` to fix it.

Oh awesome, thanks! Fixed it. Thought tutorial code examples were failing

GitBolt commented:
Is there anything left on this or its good to merge now? @acheroncrypto

acheroncrypto reviewed (approved):
Thank you for your contribution! I'm merging it now but it will take some time to show inside the app.
